### PR TITLE
feature: Support `SecretsManagerConfiguration` for Firehose Delivery Stream Snowflake Configuration

### PR DESCRIPTION
**Pull request by civitaspo**
Signed-off-by: Takahiro Nakayama <civitaspo@gmail.com>

### Description

Amazon Data Firehose supports managing secrets using SecretsManager ([`SecretsManagerConfiguration`](https://docs.aws.amazon.com/firehose/latest/APIReference/API_SecretsManagerConfiguration.html)). However, the terraform-provider-aws does not support the use of SecretsManagerConfiguration. Therefore, I have made it possible to use SecretsManagerConfiguration.

Since the implementation effort is significant, I have only implemented it within the SnowflakeConfiguration, which I need. If this implementation is accepted, it will be possible to incorporate SecretsManagerConfiguration into other configurations with separate Pull Requests.

Additionally, this change (https://github.com/hashicorp/terraform-provider-aws/pull/38151/commits/cefe73a46ff6d62cfa9a0c7dde663f34a2dae254) is not the feature I wanted to add in this Pull Request, but I found a bug while writing tests, so I fixed it as well.

### Relations

Closes https://github.com/hashicorp/terraform-provider-aws/issues/38210.
Closes https://github.com/hashicorp/terraform-provider-aws/pull/38245.
Relates https://github.com/hashicorp/terraform-provider-aws/pull/36646.

### References

- [SnowflakeDestinationConfiguration](https://docs.aws.amazon.com/firehose/latest/APIReference/API_SnowflakeDestinationConfiguration.html)
- [SecretsManagerConfiguration](https://docs.aws.amazon.com/firehose/latest/APIReference/API_SecretsManagerConfiguration.html)

### Output from Acceptance Testing

```console
% make testacc TESTS=TestAccFirehoseDeliveryStream_snowflake PKG=firehose
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.22.4 test ./internal/service/firehose/... -v -count 1 -parallel 20 -run='TestAccFirehoseDeliveryStream_snowflake'  -timeout 360m
=== RUN   TestAccFirehoseDeliveryStream_snowflakeUpdates
=== PAUSE TestAccFirehoseDeliveryStream_snowflakeUpdates
=== CONT  TestAccFirehoseDeliveryStream_snowflakeUpdates
--- PASS: TestAccFirehoseDeliveryStream_snowflakeUpdates (154.75s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/firehose   158.176s
```

### CODE REVIEW

**Comment by civitaspo** [comment]
Sorry. I intended to write the documentation after receiving a code review, without considering the maintainer's busy schedule. However, after reading the FAQ and understanding the maintainer's workload, I have also added the documentation. This Pull Request is review-ready, so I am awaiting your review.

File: website/docs/r/kinesis_firehose_delivery_stream.html.markdown [hunk 1]
```diff
@@ -793,9 +793,9 @@ The `http_endpoint_configuration` configuration block supports the following arg
 The `snowflake_configuration` configuration block supports the following arguments:
 
 * `account_url` - (Required) The URL of the Snowflake account. Format: https://[account_identifier].snowflakecomputing.com.
-* `private_key` - (Required) The private key for authentication.
-* `key_passphrase` - (Required) The passphrase for the private key.
-* `user` - (Required) The user for authentication.
+* `private_key` - (Optional) The private key for authentication. This value is required if `secrets_manager_configuration` is not provided.
+* `key_passphrase` - (Optional) The passphrase for the private key.
```
Comment (civitaspo) [inline website/docs/r/kinesis_firehose_delivery_stream.html.markdown:797]: Implementation-wise, it was originally Optional. Therefore, I will change it to the correct description.

**Comment by civitaspo** [comment]
This PR may conflict with https://github.com/hashicorp/terraform-provider-aws/pull/38245 because both make changes related to `SecretsManagerConfiguration`.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccFirehoseDeliveryStream_Splunk\|TestAccFirehoseDeliveryStream_splunk\|TestAccFirehoseDeliveryStream_Redshift\|TestAccFirehoseDeliveryStream_redshift\|TestAccFirehoseDeliveryStream_HTTPEndpoint\|TestAccFirehoseDeliveryStream_httpEndpoint\|TestAccFirehoseDeliveryStream_snowflake' PKG=firehose ACCTEST_PARALLELISM=3
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.22.5 test ./internal/service/firehose/... -v -count 1 -parallel 3  -run=TestAccFirehoseDeliveryStream_Splunk\|TestAccFirehoseDeliveryStream_splunk\|TestAccFirehoseDeliveryStream_Redshift\|TestAccFirehoseDeliveryStream_redshift\|TestAccFirehoseDeliveryStream_HTTPEndpoint\|TestAccFirehoseDeliveryStream_httpEndpoint\|TestAccFirehoseDeliveryStream_snowflake -timeout 360m
=== RUN   TestAccFirehoseDeliveryStream_redshiftUpdates
=== PAUSE TestAccFirehoseDeliveryStream_redshiftUpdates
=== RUN   TestAccFirehoseDeliveryStream_Redshift_SecretsManagerConfiguration
=== PAUSE TestAccFirehoseDeliveryStream_Redshift_SecretsManagerConfiguration
=== RUN   TestAccFirehoseDeliveryStream_snowflakeUpdates
=== PAUSE TestAccFirehoseDeliveryStream_snowflakeUpdates
=== RUN   TestAccFirehoseDeliveryStream_splunkUpdates
=== PAUSE TestAccFirehoseDeliveryStream_splunkUpdates
=== RUN   TestAccFirehoseDeliveryStream_Splunk_ErrorOutputPrefix
=== PAUSE TestAccFirehoseDeliveryStream_Splunk_ErrorOutputPrefix
=== RUN   TestAccFirehoseDeliveryStream_Splunk_SecretsManagerConfiguration
=== PAUSE TestAccFirehoseDeliveryStream_Splunk_SecretsManagerConfiguration
=== RUN   TestAccFirehoseDeliveryStream_httpEndpoint
=== PAUSE TestAccFirehoseDeliveryStream_httpEndpoint
=== RUN   TestAccFirehoseDeliveryStream_HTTPEndpoint_ErrorOutputPrefix
=== PAUSE TestAccFirehoseDeliveryStream_HTTPEndpoint_ErrorOutputPrefix
=== RUN   TestAccFirehoseDeliveryStream_HTTPEndpoint_retryDuration
=== PAUSE TestAccFirehoseDeliveryStream_HTTPEndpoint_retryDuration
=== RUN   TestAccFirehoseDeliveryStream_HTTPEndpoint_SecretsManagerConfiguration
=== PAUSE TestAccFirehoseDeliveryStream_HTTPEndpoint_SecretsManagerConfiguration
=== CONT  TestAccFirehoseDeliveryStream_redshiftUpdates
=== CONT  TestAccFirehoseDeliveryStream_Splunk_SecretsManagerConfiguration
=== CONT  TestAccFirehoseDeliveryStream_HTTPEndpoint_retryDuration
--- PASS: TestAccFirehoseDeliveryStream_Splunk_SecretsManagerConfiguration (71.24s)
=== CONT  TestAccFirehoseDeliveryStream_HTTPEndpoint_ErrorOutputPrefix
--- PASS: TestAccFirehoseDeliveryStream_HTTPEndpoint_retryDuration (71.46s)
=== CONT  TestAccFirehoseDeliveryStream_httpEndpoint
--- PASS: TestAccFirehoseDeliveryStream_httpEndpoint (82.28s)
=== CONT  TestAccFirehoseDeliveryStream_splunkUpdates
--- PASS: TestAccFirehoseDeliveryStream_HTTPEndpoint_ErrorOutputPrefix (125.67s)
=== CONT  TestAccFirehoseDeliveryStream_Splunk_ErrorOutputPrefix
--- PASS: TestAccFirehoseDeliveryStream_splunkUpdates (107.60s)
=== CONT  TestAccFirehoseDeliveryStream_HTTPEndpoint_SecretsManagerConfiguration
--- PASS: TestAccFirehoseDeliveryStream_redshiftUpdates (262.53s)
=== CONT  TestAccFirehoseDeliveryStream_snowflakeUpdates
--- PASS: TestAccFirehoseDeliveryStream_Splunk_ErrorOutputPrefix (105.05s)
=== CONT  TestAccFirehoseDeliveryStream_Redshift_SecretsManagerConfiguration
--- PASS: TestAccFirehoseDeliveryStream_HTTPEndpoint_SecretsManagerConfiguration (82.61s)
--- PASS: TestAccFirehoseDeliveryStream_snowflakeUpdates (139.41s)
--- PASS: TestAccFirehoseDeliveryStream_Redshift_SecretsManagerConfiguration (245.93s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/firehose	555.888s
```

**Comment by ewbankkit** [comment]
@civitaspo @nikhil-goenka Thanks for the contribution :tada: :clap:.

**Comment by civitaspo** [comment]
Thank you for making the changes that allow the configuration of SecretsManagerConfiguration for all services, not just Snowflake ❤️

**Pull request merged**